src/spreadinterp.cpp

```diff
@@ -918,237 +982,458 @@ void interp_cube(FLT *target, FLT *du, FLT *ker1, FLT *ker2, FLT *ker3, BIGINT i
+  static constexpr auto padding   = get_padding<FLT, 2 * ns>();
+  static constexpr auto avx_size  = batch_t::size;
+  static constexpr auto alignment = batch_t::arch_type::alignment();
   // Kernel values stored in consecutive memory. This allows us to compute
-  // values in two directions in a single kernel evaluation call.
-  FLT kernel_values[2 * MAX_NSPREAD];
-  FLT *ker1 = kernel_values;
-  FLT *ker2 = kernel_values + ns;
-  for (BIGINT i = 0; i < M; i++) { // loop over NU pts
-    FLT re0 = dd[2 * i];
-    FLT im0 = dd[2 * i + 1];
+  // values in all three directions in a single kernel evaluation call.
+  static constexpr auto ns2 = ns * FLT(0.5); // half spread width
+  std::fill(du, du + 2 * size1 * size2, 0);  // initialized to 0 due to the padding
+  for (uint64_t pt = 0; pt < M; pt++) {      // loop over NU pts
+    const auto dd_pt = initialize_complex_batch<batch_t>(dd[pt * 2], dd[pt * 2 + 1]);
     // ceil offset, hence rounding, must match that in get_subgrid...
-    BIGINT i1 = (BIGINT)std::ceil(kx[i] - ns2); // fine grid start indices
-    BIGINT i2 = (BIGINT)std::ceil(ky[i] - ns2);
-    FLT x1    = (FLT)i1 - kx[i];
-    FLT x2    = (FLT)i2 - ky[i];
-    if (opts.kerevalmeth == 0) { // faster Horner poly method
-      set_kernel_args(kernel_args, x1, opts);
-      set_kernel_args(kernel_args + ns, x2, opts);
-      evaluate_kernel_vector(kernel_values, kernel_args, opts, 2 * ns);
-    } else {
-      // due to ns being padded up to next multiple of 4 in the
-      // eval_kernel_vec_Horner and writing zeros out to this padded size, these
-      // must occur in the order x,y...
-      eval_kernel_vec_Horner(ker1, x1, ns, opts);
-      eval_kernel_vec_Horner(ker2, x2, ns, opts);
-    }
+    const auto i1 = (BIGINT)std::ceil(kx[pt] - ns2); // fine grid start indices
+    const auto i2 = (BIGINT)std::ceil(ky[pt] - ns2);
+    const auto x1 = (FLT)std::ceil(kx[pt] - ns2) - kx[pt];
+    const auto x2 = (FLT)std::ceil(ky[pt] - ns2) - ky[pt];
+    //alignas(alignment) const auto kernel_values =
+    const auto &kernel_values =
+        ker_eval<ns, kerevalmeth, FLT, batch_t>(opts, x1, x2);
+    alignas(alignment) auto *FINUFFT_RESTRICT ker1 = kernel_values.data();
+    alignas(alignment) auto *FINUFFT_RESTRICT ker2 = kernel_values.data() + MAX_NSPREAD;
     // Combine kernel with complex source value to simplify inner loop
-    FLT ker1val[2 * MAX_NSPREAD]; // here 2* is because of complex
-    for (int i = 0; i < ns; i++) {
-      ker1val[2 * i]     = re0 * ker1[i];
-      ker1val[2 * i + 1] = im0 * ker1[i];
+    // here 2* is because of complex
+    static constexpr uint8_t batches = (2 * ns + padding) / avx_size;
+    static_assert(batches > 0, "batches must be greater than 0");
+    batch_t ker1val_batches[batches];
+
+    for (uint8_t i = 0; i < (batches & ~1); i += 2) {
+      const auto ker01       = batch_t::load_aligned(ker1 + i * avx_size / 2);
+      const auto ker00       = xsimd::swizzle(ker01, zip_low_index<arch_t>);
```

Comment:
there is a confusion with naming since ker1,ker2 referred to dims (x,y) before.   But now you use ker0, ker1 for parts of avx stuff... is there a better naming eg lo & hi for avx, while keeping 1,2,3 for dimensions (x,y,z) etc?